### PR TITLE
Feature/state preparation

### PR DESCRIPTION
**Pull request by cgranade**
This PR merges the Microsoft.Quantum.Preparation improvements feature branch into `main`.

Completes #344.

Most of the code here has been reviewed already, but this PR does include some deferred action items based on feedback from @guenp.

### CODE REVIEW

File: Standard/tests/QuantumROMTests.qs [hunk 1]
```diff
@@ -43,55 +46,86 @@ namespace Microsoft.Quantum.Tests {
...
+                if (errors[i] >= IntAsDouble(3)) {
                     fail $"index {i} reconstructed coefficient incorrect. Error is {errors[i]}";
                 }
             }
         }
     }
 
-    operation QuantumROMTest() : Unit {
+    @Test("QuantumSimulator")
+    operation TestQuantumROM() : Unit {
```
Comment (guenp) [inline Standard/tests/QuantumROMTests.qs:67]:
```suggestion
    operation TestPurifiedMixedState() : Unit {
```

File: Standard/src/Preparation/Reference.qs [hunk 2]
```diff
@@ -24,17 +24,17 @@ namespace Microsoft.Quantum.Preparation {
...
         for ((leftQubit, rightQubit) in Zipped(left, right)) {
             H(leftQubit);
-            Controlled X([leftQubit], rightQubit);
+            CNOT(leftQubit, rightQubit);
         }
     }
 
     /// # Summary
-    /// Prepares the Choi–Jamiłkowski state for a given operation onto given reference
+    /// Prepares the Choi–Jamiołkowski state for a given operation onto given reference
```
Comment (guenp) [inline Standard/src/Preparation/Reference.qs:32]: Nice catch 👍

File: Standard/tests/QuantumROMTests.qs [hunk 3]
```diff
@@ -43,55 +46,86 @@ namespace Microsoft.Quantum.Tests {
...
+            within {
+                (ControlledOnInt(stateIndex + signOffset, X))(qubits! + [sign], flag);
+            } apply {
+                AssertMeasurementProbability([PauliZ], [flag], One, AbsD(expected), $"AssertSignedProbInt failed on stateIndex {stateIndex}, expected probability {expected}.", tolerance);
+            }
+        }
+    }
+
+    @Test("QuantumSimulator")
+    operation TestQuantumROMWithData() : Unit {
```
Comment (guenp) [inline Standard/tests/QuantumROMTests.qs:107]:
```suggestion
    operation TestPurifiedMixedStateWithData() : Unit {
```

File: Standard/tests/QuantumROMTests.qs [hunk 4]
```diff
@@ -13,7 +15,8 @@ namespace Microsoft.Quantum.Tests {
     open Microsoft.Quantum.Random;
 
     // Tests the discretization algorithm
-    operation QuantumROMDiscretizationTest() : Unit {
+    @Test("QuantumSimulator")
+    operation TestQuantumROMDiscretization() : Unit {
```
Comment (guenp) [inline Standard/tests/QuantumROMTests.qs:19]:
```suggestion
    operation TestPurifiedMixedStateDiscretization() : Unit {
```

File: Standard/src/Preparation/QuantumROM.qs [hunk 5]
```diff
@@ -1,4 +1,4 @@
-// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (guenp) [inline Standard/src/Preparation/QuantumROM.qs:1]: Shoud we rename this file `PurifiedMixedState.qs`?

File: Standard/src/Preparation/QuantumROM.qs [hunk 6]
```diff
@@ -10,19 +10,46 @@ namespace Microsoft.Quantum.Preparation {
     open Microsoft.Quantum.Arrays;
 
     /// # Summary
-    /// Uses the Quantum ROM technique to represent a given density matrix.
+    /// Returns an operation that prepares a a purification of a given mixed
+    /// state.
```
Comment (guenp) [inline Standard/src/Preparation/QuantumROM.qs:14]:
```suggestion
    /// Returns an operation that prepares a a purification of a given mixed state. 
    /// A "purified mixed state" refers to states of the form |ψ⟩ = Σᵢ √𝑝ᵢ |𝑖⟩ |garbageᵢ⟩ specified by a vector of 
    /// coefficients {𝑝ᵢ}. States of this form can be reduced to mixed states ρ ≔ 𝑝ᵢ |𝑖⟩⟨𝑖| by tracing over the "garbage" 
    /// register (that is, a mixed state that is diagonal in the computational basis).
    /// 
    /// See https://arxiv.org/pdf/1805.03662.pdf?page=15 for further discussion.
```

File: Standard/src/Preparation/QuantumROM.qs [hunk 7]
```diff
@@ -61,52 +87,140 @@ namespace Microsoft.Quantum.Preparation {
...
+    internal function SplitSign(coefficient : Double) : (Double, Int) {
+        return (AbsD(coefficient), coefficient < 0.0 ? 1 | 0);
     }
 
     /// # Summary
-    /// Returns the total number of qubits that must be allocated
-    /// to the operation returned by `QuantumROM`.
+    /// Returns an operation that prepares a a purification of a given mixed
+    /// state, entangled with a register representing a given collection of
+    /// data.
```
Comment (guenp) [inline Standard/src/Preparation/QuantumROM.qs:113]:
```suggestion
    /// Returns an operation that prepares a a purification of a given mixed
    /// state, entangled with a register representing a given collection of data.
    /// A "purified mixed state with data" refers to a state of the form Σᵢ √𝑝ᵢ |𝑖⟩ |𝑥ᵢ⟩ |garbageᵢ⟩, 
    /// where each 𝑥ᵢ is a bitstring encoding additional data associated with the register |𝑖⟩.
    ///
    /// See https://arxiv.org/pdf/1805.03662.pdf?page=15 for further discussion.
```

**Review by guenp — APPROVED** [review]
Just some small suggestions; otherwise looks good. Thank you and @msoeken for the great work on this!

File: Standard/src/Preparation/QuantumROM.qs (hunk 5 above)
Comment (cgranade) [reply Standard/src/Preparation/QuantumROM.qs:1]: Possibly, yes, though thankfully the filename doesn't affect the public API such that I think we can change the filename at our leisure.

File: Chemistry/tests/SamplesTests/DocsInvokingChemistryLibrary.cs [hunk 8]
```diff
@@ -19,6 +19,7 @@
 // The System namespace provides a number of useful built-in
 // types and methods that we'll use throughout this sample.
 using System;
+using System.IO;
 
 // We use this for convnience functions for manipulation arrays.
```
Comment (msoeken) [inline Chemistry/tests/SamplesTests/DocsInvokingChemistryLibrary.cs:24]:
```suggestion
// We use this for convenience functions for manipulation arrays.
```

File: Standard/src/Canon/Multiplexer.qs [hunk 9]
```diff
@@ -74,38 +75,38 @@ namespace Microsoft.Quantum.Canon {
...
                     }
                     within {
-                        X(newauxiliary[0]);
+                        X(newauxiliary);
                     } apply {
-                        _MultiplexOperationsFromGenerator(leftUnitaries, newauxiliary, newControls, target);
+                        MultiplexOperationsFromGeneratorImpl(leftUnitaries, [newauxiliary], newControls, target);
                     }
                 } else {
                     // Recursion that reduces nIndex by 1 & sets Length(auxiliary) to 1.
```
Comment (msoeken) [inline Standard/src/Canon/Multiplexer.qs:88]:
```suggestion
                    // Recursion that reduces nIndex by 1 and sets Length(auxiliary) to 1.
```

File: Standard/src/Canon/Multiplexer.qs [hunk 10]
```diff
@@ -191,63 +192,24 @@ namespace Microsoft.Quantum.Canon {
...
+        if (Length(ctrlRegister) == 0) {
+            X(target);
+        } elif (Length(ctrlRegister) == 1) {
+            CNOT(Head(ctrlRegister), target);
+        } else {
+            EqualityFactI(Length(auxRegister), Length(ctrlRegister) - 2, "Unexpected number of auxiliary qubits");
+            let controls1 = ctrlRegister[0..0] + auxRegister;
+            let controls2 = Rest(ctrlRegister);
+            let targets = auxRegister + [target];
+            ApplyToEachA(ApplyAnd, Zip3(controls1, controls2, targets));
```
Comment (msoeken) [inline Standard/src/Canon/Multiplexer.qs:212]:
```suggestion
            ApplyToEachA(ApplyAnd, Zipped3(controls1, controls2, targets));
```

File: Standard/src/Preparation/Arbitrary.qs [hunk 11]
```diff
@@ -370,7 +365,7 @@ namespace Microsoft.Quantum.Preparation {
     /// Implementation step of arbitrary state preparation procedure.
     /// # See Also
     /// - Microsoft.Quantum.Preparation.PrepareArbitraryState
-    function _StatePreparationSBMComputeCoefficients (coefficients : ComplexPolar[]) : (Double[], Double[], ComplexPolar[]) {
+    internal function StatePreparationSBMComputeCoefficients (coefficients : ComplexPolar[]) : (Double[], Double[], ComplexPolar[]) {
```
Comment (msoeken) [inline Standard/src/Preparation/Arbitrary.qs:368]:
```suggestion
    internal function StatePreparationSBMComputeCoefficients(coefficients : ComplexPolar[]) : (Double[], Double[], ComplexPolar[]) {
```

File: Standard/src/Preparation/UniformSuperposition.qs [hunk 12]
```diff
@@ -72,27 +73,24 @@ namespace Microsoft.Quantum.Preparation {
...
     /// Implementation step of <xref:microsoft.quantum.canon.prepareuniformsuperposition>
-    operation PrepareUniformSuperposition_(nIndices: Int, nQubits: Int, idxFlag: Int, qubits: Qubit[]) : Unit {
-        body (...) {
-            let targetQubits = qubits[3..3 + nQubits-1];
-            let flagQubit = qubits[0];
-            let auxillaryQubits = qubits[1..2];
-            let theta = ArcSin(Sqrt(IntAsDouble(2^nQubits)/IntAsDouble(nIndices)) * Sin(PI() / 6.0));
-            //let theta = PI() * 0.5;
+    internal operation PrepareUniformSuperpositionOracle(nIndices: Int, nQubits: Int, idxFlag: Int, qubits: Qubit[])
+    : Unit is Adj+Ctl {
```
Comment (msoeken) [inline Standard/src/Preparation/UniformSuperposition.qs:77]:
```suggestion
    : Unit is Adj + Ctl {
```

File: Standard/src/Preparation/UniformSuperposition.qs [hunk 13]
```diff
@@ -44,24 +44,25 @@ namespace Microsoft.Quantum.Preparation {
     ///     // ...
     /// }
     /// ```
-    operation PrepareUniformSuperposition(nIndices: Int, indexRegister: LittleEndian) : Unit is Adj + Ctl {
-        if(nIndices == 0) {
-            fail $"Cannot prepare uniform superposition over {nIndices} state.";
+    operation PrepareUniformSuperposition(nIndices: Int, indexRegister: LittleEndian)
+    : Unit is Adj+Ctl {
```
Comment (msoeken) [inline Standard/src/Preparation/UniformSuperposition.qs:48]:
```suggestion
    : Unit is Adj + Ctl {
```

**Review by msoeken — APPROVED** [review]
Looks good to me. Thanks! Just noticed some spelling and formatting issues.

File: Standard/src/Canon/Multiplexer.qs (hunk 10 above)
Comment (cgranade) [reply Standard/src/Canon/Multiplexer.qs:212]: Thanks for catching that one, I thought I'd gotten them all...

**Pull request merged**